pkg/policy/l4_filter_deny_test.go

```diff
@@ -0,0 +1,676 @@
+				},
+				ToPorts: []api.PortDenyRule{{
+					Ports: []api.PortProtocol{
+						{Port: "80", Protocol: api.ProtoTCP},
+					},
+				}},
+			},
+			{
+				IngressCommonRule: api.IngressCommonRule{
+					FromEndpoints: []api.EndpointSelector{api.WildcardEndpointSelector},
+				},
+				ToPorts: []api.PortDenyRule{{
+					Ports: []api.PortProtocol{
+						{Port: "80", Protocol: api.ProtoTCP},
+					},
+				}},
+			},
+		},
+	}})
+
+	buffer := new(bytes.Buffer)
+	ctx := SearchContext{To: labelsA, Trace: TRACE_VERBOSE}
+	ctx.Logging = stdlog.New(buffer, "", 0)
+
+	l4IngressDenyPolicy, err := repo.ResolveL4IngressPolicy(&ctx)
+	c.Assert(err, IsNil)
+
+	c.Log(buffer)
+
+	expected := L4PolicyMap{"80/TCP": &L4Filter{
+		Port:     80,
+		Protocol: api.ProtoTCP,
+		U8Proto:  6,
+		wildcard: wildcardCachedSelector,
+		L7Parser: "",
+		L7RulesPerSelector: L7DataMap{
+			wildcardCachedSelector: &PerSelectorPolicy{IsDeny: true},
+		},
+		Ingress:          true,
+		DerivedFromRules: labels.LabelArrayList{nil, nil},
+	}}
+
+	c.Assert(l4IngressDenyPolicy, checker.DeepEquals, expected)
+	expected.Detach(testSelectorCache)
+
+	filter, ok := l4IngressDenyPolicy["80/TCP"]
+	c.Assert(ok, Equals, true)
+	c.Assert(filter.Port, Equals, 80)
+	c.Assert(filter.Ingress, Equals, true)
+
+	c.Assert(filter.SelectsAllEndpoints(), Equals, true)
+
+	c.Assert(filter.L7Parser, Equals, ParserTypeNone)
+	c.Assert(len(filter.L7RulesPerSelector), Equals, 1)
+	l4IngressDenyPolicy.Detach(repo.GetSelectorCache())
+
+	// Case1B: implicitly deny all endpoints.
+	repo = parseAndAddRules(c, api.Rules{&api.Rule{
+		EndpointSelector: endpointSelectorA,
+		IngressDeny: []api.IngressDenyRule{
+			{
+				IngressCommonRule: api.IngressCommonRule{
+					FromEndpoints: []api.EndpointSelector{},
+				},
+				ToPorts: []api.PortDenyRule{{
+					Ports: []api.PortProtocol{
+						{Port: "80", Protocol: api.ProtoTCP},
+					},
+				}},
+			},
+			{
+				IngressCommonRule: api.IngressCommonRule{
+					FromEndpoints: []api.EndpointSelector{},
+				},
+				ToPorts: []api.PortDenyRule{{
+					Ports: []api.PortProtocol{
+						{Port: "80", Protocol: api.ProtoTCP},
+					},
+				}},
+			},
+		},
+	}})
+
+	buffer = new(bytes.Buffer)
+	ctx = SearchContext{To: labelsA, Trace: TRACE_VERBOSE}
+	ctx.Logging = stdlog.New(buffer, "", 0)
+
+	l4IngressDenyPolicy, err = repo.ResolveL4IngressPolicy(&ctx)
+	c.Assert(err, IsNil)
+
+	c.Log(buffer)
+
+	filter, ok = l4IngressDenyPolicy["80/TCP"]
+	c.Assert(ok, Equals, true)
+	c.Assert(filter.Port, Equals, 80)
+	c.Assert(filter.Ingress, Equals, true)
+
+	c.Assert(filter.SelectsAllEndpoints(), Equals, true)
+
+	c.Assert(filter.L7Parser, Equals, ParserTypeNone)
+	c.Assert(len(filter.L7RulesPerSelector), Equals, 1)
+	l4IngressDenyPolicy.Detach(repo.GetSelectorCache())
+}
+
+// Case 2: deny all at L3/L7 in one rule, and select an endpoint and deny all on L7
+// in another rule. Should resolve to just allowing all on L3/L7 (first rule
```

Comment:
`L3/L7` here should be `L3/L4` instead (no L7 in deny rules).

Reply:
Yes, I need to take a 2nd look at all test case descriptions.